### PR TITLE
Multi-thread downloading file for S3 tables

### PR DESCRIPTION
**Pull request by vdimir**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

- Multi-thread file downloading for S3 tables

### CODE REVIEW

**Comment by vdimir** [comment]
Testing parralel download with 800MB file:
```
select count(*) from s3('https://test-bucket-vdimir.s3.mds.yandex.net/test_vdimir_1/tsv/dat.tsv', 'LineAsString', 'x String')
```

<details>
<summary>Single thread</summary>

```
set max_read_buffer_size = 1048576;
set max_download_threads = 1;
```

```
1 rows in set. Elapsed: 14.128 sec. Processed 10.00 million rows, 985.81 MB (707.80 thousand rows/s., 69.78 MB/s.)
1 rows in set. Elapsed: 11.996 sec. Processed 10.00 million rows, 985.81 MB (833.58 thousand rows/s., 82.18 MB/s.)
1 rows in set. Elapsed: 11.789 sec. Processed 10.00 million rows, 985.81 MB (848.26 thousand rows/s., 83.62 MB/s.)
```

dstat:
```
-net/total-
 recv  send
 544B 1022B
  20k   41k
  46M  198k
  55M  237k
  50M  217k
  90M  374k
  62M  265k
  76M  323k
  78M  346k
  94M  397k
  85M  352k
  74M  323k
  92M  425k
  68M  303k
 538B 1994B
```
</details>



<details>
<summary>4 threads</summary>

```
set max_read_buffer_size = 1048576;
set max_download_threads = 4;
```
```
1 rows in set. Elapsed: 14.326 sec. Processed 10.00 million rows, 985.81 MB (698.06 thousand rows/s., 68.82 MB/s.)
1 rows in set. Elapsed: 14.066 sec. Processed 10.00 million rows, 985.81 MB (710.91 thousand rows/s., 70.08 MB/s.)
1 rows in set. Elapsed: 14.229 sec. Processed 10.00 million rows, 985.81 MB (702.79 thousand rows/s., 69.28 MB/s.)
```

dstat:
 ```
-net/total-
 recv  send
 544B 1393B
  46M  703k
  70M  982k
  61M  893k
  56M  818k
  55M  810k
  60M  933k
  55M  794k
  63M  962k
  37M  537k
  67M  963k
  67M  981k
  55M  789k
  52M  746k
  69M  981k
  64M  891k
1036k   14k
  18k 8072B
6249B   30k
1576B 8701B
```

</details>


<details>
<summary>4 threads with increaed buffer size (10MB)</summary>

```
set max_read_buffer_size = 10485760;
set max_download_threads = 4;
```

```
1 rows in set. Elapsed: 9.593 sec. Processed 10.00 million rows, 985.81 MB (1.04 million rows/s., 102.77 MB/s.)
1 rows in set. Elapsed: 9.609 sec. Processed 10.00 million rows, 985.81 MB (1.04 million rows/s., 102.60 MB/s.)
1 rows in set. Elapsed: 9.425 sec. Processed 10.00 million rows, 985.81 MB (1.06 million rows/s., 104.59 MB/s.)
```

dstat:
```
 544B 1020B
  42k   64k
  81M  658k
  96M  851k
  98M  837k
  99M  826k
  87M  700k
 100M  820k
  97M  903k
 100M  801k
  99M  841k
  15M   98k
 544B  913B
```

</details>



<details>
<summary>16 threads</summary>

```
set max_read_buffer_size = 1048576;
set max_download_threads = 16;
```

```
1 rows in set. Elapsed: 9.895 sec. Processed 10.00 million rows, 985.81 MB (1.01 million rows/s., 99.63 MB/s.)
1 rows in set. Elapsed: 13.236 sec. Processed 10.00 million rows, 985.81 MB (755.51 thousand rows/s., 74.48 MB/s.)
1 rows in set. Elapsed: 10.966 sec. Processed 10.00 million rows, 985.81 MB (911.90 thousand rows/s., 89.90 MB/s.)
```

dstat:
```
-net/total-
 recv  send
 544B  844B
 246B   16k
2014B 5353B
 544B 1018B
  60M 1092k
 108M 1898k
 106M 1737k
 108M 1819k
 107M 1828k
 106M 1786k
 106M 1756k
 109M 1844k
  69M 1115k
1046k   15k
1232B 4464B
 ```

</details>


Downloading same file with `axel` tool:

<details>
<summary>axel</summary>

```
axel https://test-bucket-vdimir.s3.mds.yandex.net/test_vdimir_1/tsv/dat.tsv -n1
Downloaded 863.8 Megabyte in 12 seconds. (71549.55 KB/s)
Downloaded 863.8 Megabyte in 11 seconds. (76531.71 KB/s)
```

```
axel https://test-bucket-vdimir.s3.mds.yandex.net/test_vdimir_1/tsv/dat.tsv -n4
Downloaded 863.8 Megabyte in 11 seconds. (75576.01 KB/s)
Downloaded 863.8 Megabyte in 11 seconds. (77395.39 KB/s)
```

```
axel https://test-bucket-vdimir.s3.mds.yandex.net/test_vdimir_1/tsv/dat.tsv -n16
Downloaded 863.8 Megabyte in 10 seconds. (85608.98 KB/s)
Downloaded 863.8 Megabyte in 8 seconds. (101833.45 KB/s)
```

</details>


We can see that downloading in 4 threads with 1MB buffer less optimal than signle-threaded version, due to overhead for creating separate requests, I suppose. Increasing buffer size or max threads utilize network better (according to `dsatat`).

**Review by nikitamikhaylov — APPROVED** [review]
There are some little questions in private chat

File: src/Core/Settings.h [hunk 1]
```diff
@@ -44,6 +44,8 @@ class IColumn;
     M(UInt64, max_insert_threads, 0, "The maximum number of threads to execute the INSERT SELECT query. Values 0 or 1 means that INSERT SELECT is not run in parallel. Higher values will lead to higher memory usage. Parallel INSERT SELECT has effect only if the SELECT part is run on parallel, see 'max_threads' setting.", 0) \
     M(UInt64, max_final_threads, 16, "The maximum number of threads to read from table with FINAL.", 0) \
     M(MaxThreads, max_threads, 0, "The maximum number of threads to execute the request. By default, it is determined automatically.", 0) \
+    M(MaxThreads, max_download_threads, 4, "The maximum number of threads to download data (e.g. from s3).", 0) \
+    M(UInt64, max_download_buffer_size, 10*1024*1024, "The maximal size of buffer for parallel downloading per thread (e.g. from s3).", 0) \
```
Comment (vdimir) [inline src/Core/Settings.h:48]: Not sure about default values

**Comment by UnamedRus** [comment]
Does it make sense to compare performance with https://github.com/peak/s5cmd ?

**Comment by vdimir** [comment]
I'm going to benchmark implementation again, now I use alex, but I think s5cmd can be tested too.

Implementation intended to parallel downloading one big file from s3. If there are many files (using `*`) it will be parallelized on pipeline level.

**Comment by vdimir** [comment]
I tested downloading different number of files with different settings. 
Tests performed on localhost with minio, so throughput is high.

Also some cases compared with  `s5cmd`.

Current implementation performs good for downloading one big file. Case that suffers from downloading each file is when we have many files and `max_download_threads` high enough to have 2-3 thread per each file.

<details>
<summary>Test S3 table function with different settings</summary>

```
>>> one small file (~500MB) <<<

	# time ~/bin/s5cmd --endpoint-url='http://127.0.0.1:9001' cp 's3://foo/file00000.tsv'  .
	# real    0m0.589s

	>>> 1 thread
	1.537 1.755 1.446 1.659

	>>> 4 threads, buffer 1M
	0.826 0.863 0.854 0.875

	>>> 4 threads, buffer 10M
	0.518 0.864 0.944 0.793

	>>> 16 threads, buffer 1M
	0.842 0.874 0.994 0.906

	>>> 16 threads, buffer 10M
	0.906 0.804 0.841 0.918

>>> one big file (~5GB) <<<

	# time ~/bin/s5cmd --endpoint-url='http://127.0.0.1:9001' cp 's3://foo/f10.tsv'  .
	# real    0m6.084s

	>>> 1 thread
	12.771 12.455 12.398 12.373

	>>> 4 threads, buffer 1M
	7.409 7.427 7.409 7.438

	>>> 4 threads, buffer 10M
	4.465 4.383 4.368 4.368

	>>> 16 threads, buffer 1M
	5.109 5.100 5.244 5.361

	>>> 16 threads, buffer 10M
	4.643 4.586 4.754 4.652

>>> 4 small files <<<

	>>> 1 thread
	1.971 1.927 1.963 1.982

	>>> 8 threads (2 for each file), buffer 1M
	2.267 2.203 2.256 2.262

	>>> 8 threads (2 for each file), buffer 10M
	1.701 1.650 1.682 1.676

	>>> 16 threads (4 for each file), buffer 1M
	2.097 2.012 1.978 1.970

	>>> 16 threads (4 for each file), buffer 10M
	1.545 1.595 1.511 1.532

	>>> 32 threads (8 for each file), buffer 1M
	2.182 1.977 2.022 2.078

	>>> 32 threads (8 for each file), buffer 10M
	1.526 1.462 1.532 1.504

>>> 10 small files (10x500Mb) <<<
	
	# time ~/bin/s5cmd --endpoint-url='http://127.0.0.1:9001' cp 's3://foo/file0000*.tsv'  .
	# real    0m1.643s

	>>> 1 thread
	2.989 2.970 2.965 3.036

	>>> 32 threads (3 for each file), buffer 10M
	5.031 4.965 4.987 5.034
```
</details>

**Comment by vdimir** [comment]
I removed estimation of number of objects and `max_download_threads` always used for every file not depending on `num_streams`.

Performance reevaluated with same script, almost same, case with many files and many threads per file suffers more, but it's expectable.

<details>
<summary>Details</summary>

```
>>> one small file (~500MB) <<<

	>>> 1 thread
	1.905 1.694 1.898 1.864

	>>> 4 threads, buffer 1M
	0.914 0.989 0.838 0.876

	>>> 4 threads, buffer 10M
	0.930 0.854 0.926 0.918

	>>> 16 threads, buffer 1M
	1.056 0.937 1.090 0.908

	>>> 16 threads, buffer 10M
	0.808 0.938 0.884 0.934

>>> one big file (~5GB) <<<

	>>> 1 thread
	12.538 12.185 12.392 12.216

	>>> 4 threads, buffer 1M
	7.509 7.869 7.484 7.492

	>>> 4 threads, buffer 10M
	4.513 4.557 4.573 4.483

	>>> 16 threads, buffer 1M
	6.726 6.597 6.473 6.462

	>>> 16 threads, buffer 10M
	4.779 4.692 4.634 4.730

>>> 4 small files <<<

	>>> 1 thread
	1.885 1.979 1.995 2.022

	>>> 8 threads, buffer 1M
	2.368 2.614 2.395 2.510

	>>> 8 threads, buffer 10M
	1.515 1.322 1.406 1.397

	>>> 16 threads, buffer 1M
	2.428 2.653 2.303 2.502

	>>> 16 threads, buffer 10M
	1.613 1.523 1.415 1.464

	>>> 32 threads, buffer 1M
	2.579 2.800 2.873 2.885

	>>> 32 threads, buffer 10M
	1.966 1.847 1.782 2.029

>>> 10 small files (10x500Mb) <<<

	>>> 1 thread
	2.919 2.880 2.873 2.818

	>>> 32 threads, buffer 1M
	7.411 6.510 6.324 5.985

	>>> 32 threads, buffer 10M
	3.734 3.991 4.296 4.067

```


</details>

**Comment by alexey-milovidov** [comment]
@vdimir 's comment:

> I’m not really satisfied with code (approach with separate thread pool in general) and performance is not impressive, so maybe it should be reworked, that’s why I marked it as draft.

**Comment by nikitamikhaylov** [comment]
Close because of the impressive work already done by @antonio2368

**Comment by KochetovNicolai** [comment]
#35571

**Pull request closed without merging**